### PR TITLE
Make slash command errors translatable but also work in rageshakes

### PR DESCRIPTION
**Pull request by MadLittleMods**
Make slash command errors translatable but also work in rageshakes. We want the error to be translated for the user but not in our rageshake logs.

Also updates some error messages to give more info.

Follow-up to https://github.com/matrix-org/matrix-react-sdk/pull/7372#discussion_r769556546

Depends on https://github.com/matrix-org/matrix-web-i18n/pull/3

---
This change is marked as an *internal change* (Task), so will not be included in the changelog.

Preview: https://61dd2f2fc259a9b4046f3e88--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

File: src/SlashCommands.tsx [hunk 1]
```diff
@@ -509,10 +504,16 @@ export const Commands = [
                                     useDefaultIdentityServer();
                                     return;
                                 }
-                                throw new Error(_t("Use an identity server to invite by email. Manage in Settings."));
+                                throw newTranslatableError(
+                                    "Use an identity server to invite by email. Manage in Settings.",
+                                );
                             });
                         } else {
-                            return reject(_t("Use an identity server to invite by email. Manage in Settings."));
```
Comment (MadLittleMods) [inline src/SlashCommands.tsx:515]: We probably want to update all of our `reject(_t(...))` places to get the same benefit

File: src/SlashCommands.tsx [hunk 2]
```diff
@@ -141,13 +141,25 @@ export class Command {
 
     run(roomId: string, threadId: string, args: string) {
         // if it has no runFn then its an ignored/nop command (autocomplete only) e.g `/me`
-        if (!this.runFn) return reject(_t("Command error"));
+        if (!this.runFn) {
+            return reject(
+                newTranslatableError(
+                    "Command error: This command should have been handled by another handler and we" +
+                        "  can't try to run something without a command function.",
+                ),
```
Comment (MadLittleMods) [inline src/SlashCommands.tsx:149]: Updated error string here so if it happens, there is more info to pass along

File: src/SlashCommands.tsx [hunk 3]
```diff
@@ -141,13 +141,25 @@ export class Command {
...
         const renderingType = threadId
             ? TimelineRenderingType.Thread
             : TimelineRenderingType.Room;
         if (this.renderingTypes && !this.renderingTypes?.includes(renderingType)) {
-            return reject(_t("Command error"));
+            return reject(
+                newTranslatableError(
+                    "Command error: Unable to find rendering type (%(renderingType)s)",
+                    { renderingType },
+                ),
```
Comment (MadLittleMods) [inline src/SlashCommands.tsx:161]: Updated error string here so if it ever happens, there is more info to pass along.

It seems like for these type of errors, we could be a lot more proactive in reporting them. We expect this to never happen AFAICT. But if it does, we should automatically log it to our Sentry instance and tell the user, "Sorry, it looks like you hit an error but we've just reported it to the developers to let them know." wdyt?

File: src/SlashCommands.tsx [hunk 4]
```diff
@@ -437,7 +446,11 @@ export const Commands = [
                 return success(cli.setRoomTopic(roomId, args));
             }
             const room = cli.getRoom(roomId);
-            if (!room) return reject(_t("Failed to set topic"));
+            if (!room) {
+                return reject(
+                    newTranslatableError("Failed to set topic: Unable to find room (%(roomId)s", { roomId }),
```
Comment (MadLittleMods) [inline src/SlashCommands.tsx:451]: Added more info on why something failed for the few places we check for `if (!room)`

File: src/SlashCommands.tsx (hunk 3 above)
Comment (t3chguy) [reply src/SlashCommands.tsx:161]: Sentry is opt-in in Labs, also

File: src/SlashCommands.tsx (hunk 2 above)
Comment (t3chguy) [reply src/SlashCommands.tsx:149]: this seems full of jargon, I know slash commands are power-usery but still this feels a bit much

File: src/SlashCommands.tsx (hunk 4 above)
Comment (t3chguy) [reply src/SlashCommands.tsx:451]: This is not right, this is to *view* the topic. The *set* path is the `if (args) { ... }` one above

File: src/SlashCommands.tsx [hunk 5]
```diff
@@ -141,13 +141,25 @@ export class Command {
 
     run(roomId: string, threadId: string, args: string) {
         // if it has no runFn then its an ignored/nop command (autocomplete only) e.g `/me`
-        if (!this.runFn) return reject(_t("Command error"));
+        if (!this.runFn) {
+            return reject(
+                newTranslatableError(
```
Comment (t3chguy) [inline src/SlashCommands.tsx:146]: `newTranslatableError` don't get picked up by gen-i18n, it only looks at `_t` and `_td` atm, so won't detect when we are missing en translations. We either need to use `_td` within `newTranslatableError` calls or add `newTranslatableError` to gen-i18n

Comment (t3chguy) [reply src/SlashCommands.tsx:146]: See how many missing strings there are in the i18n files now.

**Review by t3chguy — CHANGES_REQUESTED** [review]

File: src/SlashCommands.tsx (hunk 2 above)
Comment (MadLittleMods) [reply src/SlashCommands.tsx:149]: (language feedback welcome)

This is another one of the errors that users shouldn't see and I think we should just report it.

The error before was useless and ambiguous with the other `Command failed` throws to make it harder to pick out which place threw.

Comment (MadLittleMods) [reply src/SlashCommands.tsx:149]: Perhaps `Unable to handle slash command`

File: src/SlashCommands.tsx (hunk 5 above)
Comment (MadLittleMods) [reply src/SlashCommands.tsx:146]: > See how many missing strings there are in the i18n files now.

😱 Makes sense, good spot

>  or add `newTranslatableError` to gen-i18n

I think I am leaning to this as it's probably useful to others as well to have this utility for all projects,

https://github.com/matrix-org/matrix-web-i18n/blob/e5c7071e0cdf715de87ef39dc8260e11d7add2f8/scripts/gen-i18n.js#L35

Comment (MadLittleMods) [reply src/SlashCommands.tsx:146]: Is something like this workable? https://github.com/matrix-org/matrix-web-i18n/pull/3

That we can manually configure whatever translation functions we want per project as they come up.

File: src/SlashCommands.tsx (hunk 3 above)
Comment (MadLittleMods) [reply src/SlashCommands.tsx:162]: We can punt this off to another PR.

What do you think about keying off the same Labs flag `automaticErrorReporting` and only report if enabled?

File: src/SlashCommands.tsx (hunk 5 above)
Comment (t3chguy) [reply src/SlashCommands.tsx:146]: Yeah maybe, but might be strange to need to define it in matrix-react-sdk's package.json as well as element-web's package.json

File: src/i18n/strings/en_EN.json [hunk 6]
```diff
@@ -424,7 +424,6 @@
     "Advanced": "Advanced",
```
Comment (t3chguy) [inline src/i18n/strings/en_EN.json:424]: lots of strings still missing

File: src/SlashCommands.tsx (hunk 5 above)
Comment (MadLittleMods) [reply src/SlashCommands.tsx:146]: It looks like we check-in the translations for `matrix-react-sdk`. Do we recollect them from `element-web` instead of just pulling them in? When in `element-web`, does it look across all `node_modules`?

---

Thinking about similar recursive configs out there but these go from current directory inheriting outwards.

 - https://github.com/davidtheclark/cosmiconfig (what `stylelint` uses)
 - eslint: https://github.com/eslint/eslint/blob/dd58cd4afa6ced9016c091fc99a702c97a3e44f0/lib/init/npm-utils.js#L21-L40
 - jest
 
---

Otherwise, I can just hardcode `newTranslatableError` in there 🤷‍♀️ (like https://github.com/matrix-org/matrix-web-i18n/pull/2)

Comment (t3chguy) [reply src/SlashCommands.tsx:146]: element-web layer only collects its own translations which are then merged with node_modules/matrix-react-sdk/src/i18n

Comment (MadLittleMods) [reply src/SlashCommands.tsx:146]: > element-web layer only collects its own translations which are then merged with node_modules/matrix-react-sdk/src/i18n

In that case, seems reasonable to have it in both projects since it's used in both. Is this solution good enough for this to merge once we get https://github.com/matrix-org/matrix-web-i18n/pull/3 in?

If it was a transient dependency where we collect from `node_modules/matrix-react-sdk`, I would be less keen on it.

Comment (t3chguy) [reply src/SlashCommands.tsx:146]: Sure

File: src/i18n/strings/en_EN.json (hunk 6 above)
Comment (MadLittleMods) [reply src/i18n/strings/en_EN.json:424]: Strings are back 😀

`newTranslatableError` is now specified as a translatable func (see `matrix_i18n_extra_translation_funcs` in the `package.json`)

File: src/SlashCommands.tsx (hunk 5 above)
Comment (MadLittleMods) [reply src/SlashCommands.tsx:146]: https://github.com/matrix-org/matrix-web-i18n/pull/3 is now merged 🚀

`newTranslatableError` is now specified as a translatable func (see `matrix_i18n_extra_translation_funcs` in the `package.json`)

File: src/languageHandler.tsx [hunk 7]
```diff
@@ -53,9 +53,9 @@ interface ITranslatableError extends Error {
  * @param {string} message Message to translate.
  * @returns {Error} The constructed error.
  */
-export function newTranslatableError(message: string) {
+export function newTranslatableError(message: string, variables?: IVariables): ITranslatableError {
```
Comment (t3chguy) [inline src/languageHandler.tsx:56]: wondering if we should shorten this to match other funcs, maybe `_tError` or similar

**Review by t3chguy — APPROVED** [review]
LGTM otherwise

File: src/languageHandler.tsx (hunk 7 above)
Comment (MadLittleMods) [reply src/languageHandler.tsx:56]: `_tError` seems reasonable 👍 but I'll leave that taste making to you for another PR

**Comment by MadLittleMods** [comment]
Thanks for the review @t3chguy 🦓
